### PR TITLE
F/list functions

### PR DESCRIPTION
**Pull request by bazsi**
This is a continuation of #1282, which was opened against a non-master branch. It is over a couple of iterations, comments and tests, please see #1282 for more information.

### CODE REVIEW

**Comment by bazsi** [comment]
Does it help integration if I split this branch to smaller chunks? Most of these are simply new template functions that should not break existing functionality. The remaining pieces offer an improvement in kv-parser(), but that's less important than the initial couple of patches.

Thanks

File: lib/scanner/list-scanner/list-scanner.c [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+gboolean
+list_scanner_scan_next(ListScanner *self)
+{
+  g_string_truncate(self->value, 0);
+
+  while (_locate_next_non_empty_arg(self))
+    {
+      if (_parse_unquoted_empty_value(self))
+        {
+          ;
```
Comment (lbudai) [inline lib/scanner/list-scanner/list-scanner.c]: I think adding this empty statement is superfluous, the empty block statement is enough.
(not a blocking issue, just a note)

File: lib/scanner/list-scanner/tests/test_list_scanner.c [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+  LIST_SCANNER_TESTCASE(test_list_scanner_unquoted_empty_items_are_skipped_to_make_it_easy_to_concatenate_lists);
+  LIST_SCANNER_TESTCASE(test_list_scanner_quoted_empty_items_are_parsed_as_empty_values);
+  LIST_SCANNER_TESTCASE(test_list_scanner_comma_delimiter_values_are_split);
+  LIST_SCANNER_TESTCASE(test_list_scanner_comma_and_arg_are_equivalent);
+  LIST_SCANNER_TESTCASE(test_list_scanner_works_with_gstring_input);
+  LIST_SCANNER_TESTCASE(test_list_scanner_handles_single_quotes);
+  LIST_SCANNER_TESTCASE(test_list_scanner_handles_double_quotes);
+  LIST_SCANNER_TESTCASE(test_list_scanner_malformed_quotes);
+  return 0;
+}
```
Comment (lbudai) [inline lib/scanner/list-scanner/tests/test_list_scanner.c]: For new tests we prefer to use the criterion framework.

File: modules/basicfuncs/list-funcs.c [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+_adjust_list_slice(gint argc, GString *argv[],
+                   gint *first_ndx, gint *last_ndx)
+{
+  gint count = -1;
+
+  if (*first_ndx < 0 || *last_ndx < 0)
+    count = _list_count(argc, argv);
+
+  if (*first_ndx < 0)
+    *first_ndx += count;
```
Comment (lbudai) [inline modules/basicfuncs/list-funcs.c]: `first_ndx` can still be a negative number as I see

Comment (lbudai) [inline modules/basicfuncs/list-funcs.c]: Or `parse_number` accepts only non-negative values and when the user sets a value that cannot be parse as a number then it sets the output parameter value to -1? (I didn't check it)

File: modules/basicfuncs/context-funcs.c [hunk 4]
```diff
@@ -43,3 +43,47 @@ tf_context_length_free_state(gpointer s)
...
+void
+tf_context_lookup_call(LogTemplateFunction *self, gpointer s, const LogTemplateInvokeArgs *args, GString *result)
+{
+  gint i, msg_ndx;
+  gboolean first = TRUE;
+  TFCondState *state = (TFCondState *) s;
+  gint count = 0;
+  GString *buf = g_string_sized_new(64);
+
+  for (msg_ndx = 0; msg_ndx < args->num_messages; msg_ndx++)
```
Comment (lbudai) [inline modules/basicfuncs/context-funcs.c]: I'd limit the scope of `msg_ndx` to the for cycle. (just a note)

**Comment by kira-syslogng** [comment]
Can one of the admins verify this patch?

**Comment by pzoleex** [comment]
retest this please

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: modules/basicfuncs/list-funcs.c (hunk 3 above)
Comment (bazsi) [inline modules/basicfuncs/list-funcs.c]: parse_number() can parse negative numbers, it does parsing with strtoll(). 

The negative values are actually supported values in slice boundaries and are interpreted similarly to Python does that.

```
$(list-slice -1: $LIST) - is the last element (negative elements index from the end of the list)
$(list-slice :-1 $LIST) - is until the last element
```

_adjust only adds "count", which might still be negative, but as I see those cases are handled in the caller. as the loop to skip initial entries would not be entered, thus we would be using entries from the very first element.

```
$(list-slice -100: a,b,c) - returns "a,b,c"
$(list-slice -1: a,b,c) - returns "c"
```
Similarly for the end index. If it is too low, the 'slicing' loop will not be entered (the 2nd loop in _list_slice), causing no elements of the list being returned.

Anyway, adding a comment to make this clear.

File: lib/scanner/list-scanner/list-scanner.c (hunk 1 above)
Comment (bazsi) [inline lib/scanner/list-scanner/list-scanner.c]: I have refactored this code to make this more readable.

**Comment by bazsi** [comment]
I have fixed the notes with a couple of followup patches, please let me know if you are satisfied.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: modules/basicfuncs/list-funcs.c (hunk 3 above)
Comment (lbudai) [inline modules/basicfuncs/list-funcs.c]: thanks for the info

**Comment by bazsi** [comment]
\o/

**Pull request merged**